goal_src/jak1/engine/math/math.gc

```diff
@@ -317,3 +317,67 @@
   `(+ -1.0 (the-as float (logior #x3f800000 (/ (rand-uint31-gen ,gen) 256))))
   )
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; mimicking Java's util.Random
+;;;; https://docs.oracle.com/javase/8/docs/api/java/util/Random.html
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+(define *java-rand-seed* (the uint64 42069))
+
+(defun java-rand-init ((seed uint))
+  (set! *java-rand-seed* seed)
+)
+
+(defun java-rand-advance-seed ()
+  ;; seed =  (seed * 0x5DEECE66DL + 0xBL) & ((1L << 48) - 1)
+  (set! *java-rand-seed* (the uint64 (logand (+ (* *java-rand-seed* 25214903917) 11) (- (shl 1 48) 1))))
+)
+
+(defun java-rand-next ((bits int))
+  ;; TODO: figure out why we need to offset bits like this, probably getting converted to uint32 somewhere or something
+  (+! bits 16) 
```

Comment:
Need some help debugging this, seemed like uint64s were getting downgraded to uint32s possibly, and then the `(shr *java-rand-seed* (- 48 bits))` was not working as it does in the Java library